### PR TITLE
Feature: Export Library

### PR DESCRIPTION
**Pull request by jcharkow**
This adds functionality to export a .oswpq file or a .oswpqd file to a library. The library can use either experimental or the previous libraries RT/IM or fragment ion intensity. 

Currently .osw and .parquet are unsupported but can be added in the future.

### CODE REVIEW

File: pyprophet/io/_base.py [hunk 1]
```diff
@@ -619,6 +619,61 @@ def export_results(self, data: pd.DataFrame):
...
+        data = data.sort_values(by=['Q_Value', 'RunId']).groupby("TransitionId").head(1)
+        assert (len(data['TransitionId'].drop_duplicates()) == len(data))
+
+        # Remove Annotation Column if all NAN
+        if data['Annotation'].isnull().all() or data['Annotation'].eq("NA").all():
+            logger.debug("Annotation column is empty, so computing it manually.")
+            data.drop(columns=['Annotation'], inplace=True)
+            data['Annotation'] = data['FragmentType'] + data['FragmentSeriesNumber'].astype(str) + '^' + data['FragmentCharge'].astype(str)
+
+        import sklearn.preprocessing as preprocessing
```
Comment (singjc) [inline pyprophet/io/_base.py:643]: Can you move this import to the top with the others above?

File: pyprophet/io/_base.py [hunk 2]
```diff
@@ -619,6 +619,61 @@ def export_results(self, data: pd.DataFrame):
...
+        This function cleans the original dataframe and exports the library
+
+        Args:
+            data: Input DataFrame with library data
+
+        """
+        cfg = self.config
+
+        # For precursors found in more than one run, select the run with the smallest q value
+        # If q values are the same, select the first run
```
Comment (singjc) [inline pyprophet/io/_base.py:633]: Would it make a difference if we sort by intensity as well? So if there are ties in peakgroup q-values, then we select the highest intensity transition, and lastly if there are ties in both q-values and intensity, then select the first entry. Or would it not make too much of a difference?

File: pyprophet/io/_base.py [hunk 3]
```diff
@@ -619,6 +619,61 @@ def export_results(self, data: pd.DataFrame):
...
+        data = data.sort_values(by=['Q_Value', 'RunId']).groupby("TransitionId").head(1)
+        assert (len(data['TransitionId'].drop_duplicates()) == len(data))
```
Comment (singjc) [inline pyprophet/io/_base.py:635]: Can you add an assertion message?

File: pyprophet/io/_base.py [hunk 4]
```diff
@@ -619,6 +619,61 @@ def export_results(self, data: pd.DataFrame):
...
+
+        import sklearn.preprocessing as preprocessing
+        if cfg.rt_calibration and cfg.rt_unit == "iRT":
+            data['NormalizedRetentionTime'] = preprocessing.MinMaxScaler().fit_transform(data[['NormalizedRetentionTime']]) * 100
+        if cfg.intensity_calibration:
+            data['LibraryIntensity'] = (
+            data['LibraryIntensity'] /
+            data.groupby('Precursor')['LibraryIntensity'].transform('max') *
+            10000)
+            data = data[data['LibraryIntensity'] > 0] # Remove rows with zero intensity
```
Comment (singjc) [inline pyprophet/io/_base.py:651]: Not sure if it would be useful information to add a info/debug log for how many entries are 0 that get removed?

File: pyprophet/io/export/osw.py [hunk 5]
```diff
@@ -123,6 +123,9 @@ def _read_sqlite(self, con):
         """Main entry point for reading SQLite data, delegates to specific methods."""
         cfg = self.config
 
+        if self.config.export_format == "library":
+            raise NotImplementedError("Library export from non-split .parquet files is not supported")
```
Comment (singjc) [inline pyprophet/io/export/osw.py:127]:
```suggestion
            raise NotImplementedError("Library export from sqlite OSW files is not supported")
```

File: pyprophet/cli/export.py [hunk 6]
```diff
@@ -347,6 +348,121 @@ def export_matrix(
...
+    default=0.01,
+    show_default=True,
+    type=float,
+    help="Filter results to maximum global protein-level q-value, should not use values > 0.01.",
+)
+@click.option(
+    "--rt_calibration/--no-rt_calibration",
+    default=True,
+    show_default=True,
+    help="Use empirical RT values as oppose to the original library RT values."
```
[the remark this answers is not in the document]
Comment (singjc) [reply pyprophet/cli/export.py:392]: I guess the `rt_unit` is added to control for whether experimental RT values are returned as raw values or normalized?

**Review by singjc — CHANGES_REQUESTED** [review]
Thanks for the addition! Looks mostly good to go, I just had a few questions and suggestions.

**Review by singjc — APPROVED** [review]
Great, thanks for the changes. Will merge.
